### PR TITLE
MaybeSource rewritten as GraphStage 

### LINKED ISSUE #22789: Rewrite MaybeSource as a GraphStage
The surprise when I noticed that it is a `SourceModule`: 7/10

### PR DESCRIPTION
**Pull request by johanandren**
Fixes #22789, includes a FailedSource that is a GraphStage as well.

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/ActorMaterializerSpec.scala [hunk 1]
```diff
@@ -24,7 +24,7 @@ class ActorMaterializerSpec extends StreamSpec with ImplicitSender {
     "properly shut down actors associated with it" in {
       val m = ActorMaterializer.create(system)
 
-      val f = Source.maybe[Int].runFold(0)(_ + _)(m)
+      val f = Source.fromPublisher(TestPublisher.probe[Int]()(system)).runFold(0)(_ + _)(m)
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/ActorMaterializerSpec.scala]: Wanted to keep the boundary here.

**Comment by johanandren** [comment]
mima...

**Review by patriknw — APPROVED** [review]
LGTM

File: project/MiMa.scala [hunk 2]
```diff
@@ -1222,14 +1220,26 @@ object MiMa extends AutoPlugin {
...
         ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.cluster.ddata.DeltaPropagationSelector.maxDeltaSize"),
-        
+
         // #23023 added a new overload with implementation to trait, so old transport implementations compiled against
         // older versions will be missing the method. We accept that incompatibility for now.
         ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.remote.transport.AssociationHandle.disassociate")
+      ),
+      "2.5.3" -> Seq(
+        // #22789 Source.maybe rewritten as a graph stage
+        ProblemFilters.exclude[MissingClassProblem]("akka.stream.impl.MaybePublisher"),
```
Comment (raboof) [inline project/MiMa.scala]: Could `ProblemFilters.excludePackage("akka.stream.impl")` work here? (https://github.com/typesafehub/migration-manager/issues/171)

Comment (johanandren) [inline project/MiMa.scala]: ooh, nice, hadn't seen that one!

File: akka-stream/src/main/scala/akka/stream/impl/MaybeSource.scala [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+@InternalApi private[akka] object MaybeSource extends GraphStageWithMaterializedValue[SourceShape[AnyRef], Promise[Option[AnyRef]]] {
+  val out = Outlet[AnyRef]("MaybeSource.out")
+  override val shape = SourceShape(out)
+
+  override protected def initialAttributes = DefaultAttributes.maybeSource
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Promise[Option[AnyRef]]) = {
+    import scala.util.{ Success ⇒ ScalaSuccess, Failure ⇒ ScalaFailure }
+    val promise = Promise[Option[AnyRef]]()
+    val logic = new GraphStageLogic(shape) with OutHandler {
```
Comment (raboof) [inline akka-stream/src/main/scala/akka/stream/impl/MaybeSource.scala]: Might be neat to make this an inner class that only takes the `Future` and does not have access to the `Promise` - but that's nitpicking.

**Review by raboof — APPROVED** [review]
Neat, splitting it out to separate files also cleans things up nicely.

**Pull request merged**